dot/state/grandpa.go

```diff
@@ -24,44 +26,570 @@ var (
+		return false, nil
+	}
+
+	pendingChangeNode := &pendingChangeNode{change: pendingChange, nodes: []*pendingChangeNode{}}
+	c.nodes = append(c.nodes, pendingChangeNode)
+	return true, nil
+}
+
+type orderedPendingChanges []*pendingChange
+
+func (o orderedPendingChanges) Len() int      { return len(o) }
+func (o orderedPendingChanges) Swap(i, j int) { o[i], o[j] = o[j], o[i] }
+
+// Less order by effective number and then by block number
+func (o orderedPendingChanges) Less(i, j int) bool {
+	return o[i].effectiveNumber() < o[j].effectiveNumber() &&
+		o[i].announcingHeader.Number < o[j].announcingHeader.Number
+}
+
 // GrandpaState tracks information related to grandpa
 type GrandpaState struct {
-	db chaindb.Database
+	db         chaindb.Database
+	blockState *BlockState
+
+	forksLock sync.RWMutex
+
+	scheduledChangeRoots []*pendingChangeNode
+	forcedChanges        orderedPendingChanges
 }
 
 // NewGrandpaStateFromGenesis returns a new GrandpaState given the grandpa genesis authorities
-func NewGrandpaStateFromGenesis(db chaindb.Database, genesisAuthorities []types.GrandpaVoter) (*GrandpaState, error) {
+func NewGrandpaStateFromGenesis(db chaindb.Database, bs *BlockState,
+	genesisAuthorities []types.GrandpaVoter) (*GrandpaState, error) {
 	grandpaDB := chaindb.NewTable(db, grandpaPrefix)
 	s := &GrandpaState{
-		db: grandpaDB,
+		db:         grandpaDB,
+		blockState: bs,
 	}
 
 	if err := s.setCurrentSetID(genesisSetID); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot set current set id: %w", err)
 	}
 
 	if err := s.SetLatestRound(0); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot set latest round: %w", err)
 	}
 
 	if err := s.setAuthorities(genesisSetID, genesisAuthorities); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot set authorities: %w", err)
 	}
 
-	if err := s.setSetIDChangeAtBlock(genesisSetID, 0); err != nil {
-		return nil, err
+	if err := s.setChangeSetIDAtBlock(genesisSetID, 0); err != nil {
+		return nil, fmt.Errorf("cannot set change set id at block 0: %w", err)
 	}
 
 	return s, nil
 }
 
 // NewGrandpaState returns a new GrandpaState
-func NewGrandpaState(db chaindb.Database) (*GrandpaState, error) {
+func NewGrandpaState(db chaindb.Database, bs *BlockState) (*GrandpaState, error) {
 	return &GrandpaState{
-		db: chaindb.NewTable(db, grandpaPrefix),
+		db:         chaindb.NewTable(db, grandpaPrefix),
+		blockState: bs,
 	}, nil
 }
 
+// HandleGRANDPADigest receives a decoded GRANDPA digest and calls the right function to handles the digest
+func (s *GrandpaState) HandleGRANDPADigest(header *types.Header, digest scale.VaryingDataType) error {
+	switch val := digest.Value().(type) {
+	case types.GrandpaScheduledChange:
+		return s.addScheduledChange(header, val)
+	case types.GrandpaForcedChange:
+		return s.addForcedChange(header, val)
+	case types.GrandpaOnDisabled:
+		return nil
+	case types.GrandpaPause:
+		return nil
+	case types.GrandpaResume:
+		return nil
```

Comment:
should we error for these too (as we are probably supposed to support them?)

Reply:
Yes, I'm fixing the tests cases for the `scheduled` and `forced` changes before implement the `pause` and `resume` digests